### DISCUSSION TITLE
Building off pre existing yolov8n-pose.pt

### DISCUSSION THREAD

**Discussion by StatisticalHorse** [ultralytics/ultralytics#25604] [Q&A]
As the title says, I need to build another model that adds more classes to the nano pose model (and keeps the human pose recognition that it already has). I am trying to see if there is a way to do so, as when I _did_ have a custom human recognition in my dataset it was projecting all the key points onto one side. I would prefer not to have to run 2 separate models for what I am working on. Thanks!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @StatisticalHorse, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many common questions may already be answered.

For this custom pose-training ❓ question, please provide as much information as possible, including your dataset format, class and keypoint definitions, sample images with annotations, model configuration, training command, and training logs. Please also verify that your workflow follows our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips). If you are encountering unexpected keypoint behavior, please include a 🐛 [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) so we can investigate it effectively.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

This is an automated response 🤖, and an Ultralytics engineer will also assist you soon.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Fine-tuning `yolov8n-pose.pt` only on custom data will not reliably preserve COCO human pose; train on a combined dataset containing the original 17-keypoint person annotations plus your new classes. Pose models use one global `kpt_shape`, so all classes must share that keypoint schema; otherwise use separate models. Keypoints collapsing to one side usually indicates incorrect normalized coordinates, keypoint ordering, or `flip_idx`; verify these against the [pose dataset format](https://docs.ultralytics.com/datasets/pose), then delete any `*.cache` files under `labels/` before retraining.

**Comment by fadyabirached** [comment]
You can do this in one model, don't need two. The thing to understand is YOLOv8-pose has one keypoint skeleton (kpt_shape) for the whole model, it's not per class.

So the fix isn't training separate pose heads per class. Build your dataset so non-human classes have their keypoints marked invisible.

Keep kpt_shape: [17, 3] in your data.yaml, don't touch it. For your new non-human classes, still include 17 keypoint slots per label, but set visibility to 0 for all of them, x/y can just be 0 too. That tells the loss function to skip keypoint loss for those instances, they just get bbox + class loss.

Fine-tune starting from yolov8n-pose.pt, not from scratch, so you keep the pretrained pose head for the human class and just extend nc for your new classes.

That "projecting keypoints onto one side" thing you saw before sounds like exactly this problem. If your non-human instances had visibility not set to 0, or had garbage keypoint values instead of properly zeroed and invisible ones, the model tries to learn keypoint positions for objects with no consistent keypoint geometry, and it drags everything toward some averaged, degenerate solution.

So: single model, single kpt_shape, mixed dataset where only the person class has real keypoint annotations and everything else has them zeroed and marked invisible. No need for two models.

**Reply by glenn-jocher** [reply]
Yes—one model can handle this, but `kpt_shape` is global rather than per class. Keep `kpt_shape: [17, 3]`, annotate people with the 17 COCO keypoints, and provide 17 `(0, 0, 0)` keypoints for each non-person instance so their coordinate loss is masked; fine-tune from `yolov8n-pose.pt` while retaining representative person-pose training data to avoid forgetting. See the [pose dataset format](https://docs.ultralytics.com/datasets/pose), delete any `*.cache` files under `labels/` after correcting annotations, and filter non-person keypoints during inference.

**Comment by rishig47-dev** [comment]
If you want to keep the existing human-pose estimation and add more information, I would first clarify whether the new outputs are **object classes** or **additional keypoints**.

If they are additional body/keypoint locations, you should extend the pose dataset's `kpt_shape` and annotations rather than adding them as detection classes. For example, YOLOv8-pose uses `kpt_shape: [17, 3]` for 17 keypoints with `(x, y, visibility)`, and Ultralytics reads the `kpt_shape` directly from the dataset configuration.

You can still start from the pretrained `yolov8n-pose.pt` weights. Ultralytics transfers compatible pretrained weights when fine-tuning on a custom pose dataset, while layers whose shapes no longer match are reinitialized.

So I would recommend:

1. Keep the existing pose keypoints you want to retain.
2. Add the new keypoints to `kpt_shape` and annotate them consistently in every relevant image.
3. Update `flip_idx` if you have symmetric keypoints.
4. Train using the pretrained `yolov8n-pose.pt` as the starting point.
5. Validate the new keypoints separately to make sure the additional outputs are actually learned.

If by "more classes" you mean completely different objects rather than additional keypoints, then that is a different setup: you would change the dataset `names`/`nc` and train the appropriate detection/pose configuration.

**Reply by pderrenger** [reply]
Yes, but `yolov8n-pose.pt` is only the initialization: train on a combined dataset containing the original `person` class with COCO’s exact 17-keypoint order plus your new classes, keeping one global `kpt_shape: [17, 3]`; use visibility `0` for unlabeled keypoints on classes without poses. Keypoints collapsing to one side usually indicates incorrect normalization, ordering, or `flip_idx`; compare labels with the [pose dataset format](https://docs.ultralytics.com/datasets/pose), inspect the generated training plots, then delete `labels/*.cache` after correcting them.

**Comment by StatisticalHorse** [comment]
Sorry for taking so long to respond. I just want it plain: is there any way to simply _add on_ to the pretrained model with other non-human classes while keeping the integrity and responsiveness the pretrained YOLOv8n-pose model? Thank you.

**Reply by pderrenger** [reply]
No—not without retraining. Adding classes changes the pose model’s classification head, and fine-tuning only on new classes can degrade the pretrained person capability. For one standard model, train `yolov8n-pose.pt` on a combined dataset containing the original human-pose examples plus your new classes, using one shared `kpt_shape`, then validate both tasks; retaining the original accuracy and latency is not guaranteed. See the [pose dataset format](https://docs.ultralytics.com/datasets/pose).

**Comment by navalkishoremehta95** [comment]
You can keep the COCO person pose head and add other classes, but the dataset has to stay consistent with how pose models are trained.

**What usually goes wrong**  
If you mix “person + keypoints” with new classes that have **no keypoints** (or wrong keypoint layout), the pose branch gets noisy and keypoints collapse to one side. Also, `kpt_shape` / skeleton must match the pretrained `yolov8n-pose` layout (COCO-17) for the person class.

**Practical approach**
1. Start from `yolo pose train model=yolov8n-pose.pt data=your.yaml`.
2. In `data.yaml`, keep **person** as a class with the same 17 keypoints.
3. Add new classes only if their labels are valid for the pose format you use (bbox + optional keypoints). For pure detection classes without keypoints, many people instead run **pose + detect** as two heads/models, or use a custom multi-task setup — a single stock pose checkpoint is not ideal for “person pose + arbitrary object classes” unless every class has a defined keypoint schema.
4. Make sure every person instance has keypoints in the label files; empty/malformed kpts are a common cause of one-sided projections.

**If you only need human pose + a few objects**  
Often cleaner: keep `yolov8n-pose` for people, and a small `yolov8n` detect model for the extra classes, then fuse results in post-process. One model is nicer operationally, but two small models are usually more stable than fighting the pose label format.

If you paste your `data.yaml` and one sample label line for person vs a new class, we can check whether the format matches pose training.

**Reply by pderrenger** [reply]
A single pose model can detect extra classes, but it uses one global `kpt_shape`; keep COCO’s `kpt_shape: [17, 3]`, include COCO-style person examples to prevent forgetting, and give non-pose objects 17 zeroed `(0, 0, 0)` keypoints so pose loss is masked. One-sided poses usually indicate incorrect keypoint order, visibility values, or `flip_idx`; verify them against the [pose dataset format](https://docs.ultralytics.com/datasets/pose), then delete any `*.cache` files under `labels` before retraining from `yolov8n-pose.pt`.

**Reply by StatisticalHorse** [reply]
Thank you! I am currently developing my project with 2 models, but I am trying to design a mobile app, and am getting almost 2 frames inferences per second on a good day. I am trying to switch it to a one-model setup, and will try your advice for building it, though if you have a better solution for in-service inference, I would be happy to hear and take it into account.

**Reply by pderrenger** [reply]
Before retraining, profile an optimized mobile export: use [CoreML](https://docs.ultralytics.com/integrations/coreml) on iOS or [LiteRT](https://docs.ultralytics.com/integrations/litert) on Android with hardware acceleration and quantization, then test smaller `imgsz` values such as `320` or `416`. If two models remain necessary, run pose every frame and the object detector less frequently; otherwise, the combined pose model described above is the cleaner single-pass option.
